### PR TITLE
[WIP] support register_hook on intermediate tensors

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107502

The strategy in this PR is pretty straightforward.

A few simple rules and assumptions:

1) We do not want `register_hook` calls on input tensors in the captured graph. 
2) We also do not want to graph break on `register_hook`.
3) aot_autograd will inline the op contents of bwd hooks.
4) We are alright with specializing on hook function identity.
5) We do not support lambda backward hooks and will graph break on them.
6) We do not dynamo trace the contents of backward hooks - this means that they operate a bit like `allow_in_graph` in the sense that aot_autograd will trace away the contents of the hook. As this is already the case for how backward_hooks behave w/o dynamo interceding, we do not care about this, but it is a case worth noting. 

So, the plan is to deal with 2 kinds of hooks:
1) Hooks on objects with sources (inputs, params)
2) Hooks on objects w/o sources (intermediaries, and outputs). 

Note: As outputs can be made simple by how dynamo handles residuals, they could actually be handled as if they were inputs, but, for the sake of this PR, we will refer to hooks as either hooks on inputs (sourced), or hooks on intermediaries (not sourced).

The plan:

For tensors w/ a source:
We record registered hooks, store them as a global, and associate them with the tensor in residuals. This means that when dynamo goes to create the frame, where we produce bytecode to stitch together our PT2 modified bytecode with the original eager code, we call `register_hook`. This registration of hooks in residuals is sound because (a) it happens right after a Pt2 frame region ends and (b) we know that the tensor is alive in f_locals, f_globals, or a module in the users invoking frame. This means we can soundly know it will be around to invoke `register_hook` on. As long as we guard on the identity of the lifted function, this is sound to do. 

For tensors w/o a source:
We do the same lifting and storing into a global as we do for hooks on a tensor w/ a source - but instead of registering hooks in residuals, we lift the hook arg to input. This gives it a placeholder proxy, which allows us to trace the register_hook call into the graph, and preserve it until it get invoked and traced away in aot_autograd.

Handles:

An interesting new component here is the creation of a `STORE_FAST `->`LOAD_FAST` associated with the handle, the return result of `register_hook`. If the user code stored the result of `register_hook` in a handle, we need to honor that. We do so by interceding into `STORE_FAST`, and recording the name of the local variable as directed by user code. We then honor that same name in the reconstructed bytecode. If the user did not store a hook, we merely pop the produced value to preserve the stack.

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

File: torch/_dynamo/side_effects.py [hunk 1]
```diff
@@ -375,6 +380,39 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            # We do not record register_hook in the graph, as we do not have any mechanisms for storing the func
+            # associated into the graph. Instead, we track them and record them here, in residuals.
+            # register_hook does not graph break, but removing the handle does. First, any function passed to register
+            # is lifted into a global variable. At the end of codegen, here, in side_effects, we iterate over all the
+            # tensors with stored hooks, and we codegen
+            # (1) the tensor
+            # (2) a register_hook call on the tensor, registering the global function lifted and recorded above
+            # (3) a handle if one was created in eager
+            # Note - the handle bit is tricky, because we need to get the real user variable name it is referenced as.
+            # This real name is stored as last_seen_name, and associated where we handle STORE_FAST
```
Comment (voznesenskym) [inline torch/_dynamo/side_effects.py:399]: rewrite

**Comment by voznesenskym** [comment]
This is WIP - there's a few cases and things I need to change

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -582,3 +583,19 @@ def __init__(self, value, **kwargs):
...
+        self, value, last_seen_name=None, value_type=None, mutable_local=None, **kwargs
+    ):
+        super().__init__(value, value_type, **kwargs)
+        # associated later, see code symbolic_convert and On dynamo tensor_hooks
+        self.last_seen_name = last_seen_name
+        self.mutable_local = mutable_local
+
+    def reconstruct(self, codegen):
+        if self.last_seen_name:
+            # It is an invariant that at this point, a STORE_FAST was executed for this name.
```
Comment (voznesenskym) [inline torch/_dynamo/variables/user_defined.py:599]: Self notes - this invariant can actually be enforced a little better

File: torch/_dynamo/variables/user_defined.py [hunk 3]
```diff
@@ -582,3 +583,19 @@ def __init__(self, value, **kwargs):
...
+            return [create_instruction("LOAD_FAST", argval=self.last_seen_name)]
```
Comment (voznesenskym) [inline torch/_dynamo/variables/user_defined.py:600]: Self notes - A LOAD_DEREF considering util exists, we can use that here

File: torch/_dynamo/symbolic_convert.py [hunk 4]
```diff
@@ -784,7 +788,14 @@ def LOAD_DEREF(self, inst):
         self.push(self.symbolic_locals[inst.argval])
 
     def STORE_FAST(self, inst):
-        self.symbolic_locals[inst.argval] = self.pop()
+        loaded_vt = self.pop()
+        name = inst.argval
+        # If the last top VT in the stack (just popped) is a handle (see On dynamo tensor_hooks), we associate
+        # the stored name.
+        if isinstance(loaded_vt, RemovableHandleVariable):
+            new_rhv = loaded_vt.clone(last_seen_name=name)
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:796]: Self notes - this should add a new var to co_varnames, ensuring a unique name and proper population for certain cases

File: torch/_dynamo/side_effects.py [hunk 5]
```diff
@@ -375,6 +380,45 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            #   - Incorporate a handle if one was established in the eager phase.
+            # The handle's exact user-specified name, "last_seen_name", is discerned and associated during STORE_FAST.
+            cg(tensor)
+            cg.extend_output([cg.create_load_attr("register_hook")])
+            cg(hook)
+            cg.extend_output(create_call_function(1, True))
+            if hasattr(handle, "last_seen_name") and handle.last_seen_name:
+                # register_hook stored with variable name assigned to the handle
+                cg.extend_output(
+                    [create_instruction("STORE_FAST", argval=handle.last_seen_name)]
```
Comment (voznesenskym) [inline torch/_dynamo/side_effects.py:413]: Self notes - a STORE_DEREF considering util exists, use it

File: torch/_dynamo/side_effects.py [hunk 6]
```diff
@@ -375,6 +380,45 @@ def codegen_save_tempvars(self, cg: PyCodegen):
                 ]
             )
 
+        for (
+            tensor,
+            hook,
+            handle,
+        ) in self.tensor_hooks:
```
Comment (voznesenskym) [inline torch/_dynamo/side_effects.py:387]: Self notes - Move to its own util, invoke as the first thing in pass 1, pass 2

File: torch/_dynamo/side_effects.py [hunk 7]
```diff
@@ -375,6 +380,54 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            #
+            # Dynamo Tensor Hooks Workflow:
+            # - Functions passed to register_hook are lifted globally.
+            # - In the "side_effects" phase of codegen, we iterate over tensors with hooks to:
+            #   - Generate the tensor.
+            #   - Issue a register_hook call on the tensor, linking to the globally stored function.
+            #   - Incorporate a handle if one was established in the eager phase.
+            # The handle's exact user-specified name, "last_seen_name", is discerned and associated during STORE_FAST.
+            # return
+            cg(tensor)
```
Comment (voznesenskym) [inline torch/_dynamo/side_effects.py:416]: Self note - may need a different load() mechanism, can cause re-entrancy bugs here. Should be easy to pass in a load fn for output cases where tensor is already loaded.

**Comment by albanD** [comment]
Does this mean that each Tensor on which register_hook is called has to be promoted to be an output of the graph so that you can register the hook? Is that ok?
Also how does AOTAutograd interract with this? Will these hooks be properly registered for AOTAutograd to be able to trace through them?

**Comment by voznesenskym** [comment]
Quick update - this is not so far off, and we will reuse most of this PR - but will add special handling for intermediate hooks as proven out by our aot_autograd + FSDP workstream here https://fb.workplace.com/groups/257735836456307/permalink/553768123519742/

File: torch/_dynamo/output_graph.py [hunk 8]
```diff
@@ -1138,7 +1140,8 @@ def remove_unused(node):
                     fake = (
                         arg.fake_tensor if arg.fake_tensor is not None else arg.example
                     )
-                    used_symbols |= free_symbols(fake)
+                    if arg.has_symbols:
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:1143]: How come? If it doesn't have free symbols, free_symbols returns empty set which is fine too

File: torch/_functorch/aot_autograd.py [hunk 9]
```diff
@@ -742,7 +744,10 @@ def from_fun(t):
     @wraps(f)
     def inner(*flat_args):
         # This function is meant to be run with the forward, which expects a flat list of tensor/symint/other args.
-        assert all(isinstance(a, KNOWN_TYPES) for a in flat_args)
+        def valid_input(a):
+            # Insanely sus, talk to folks about this
+            return isinstance(a, KNOWN_TYPES) or callable(flat_args[-1])
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:749]: Yeah I guess I still really don't understand why we want hook functions to be explicit arguments to the FX graph. I would almost prefer it if there was some extralingual mechanism for getting at them. @jansel could you walk me through the reasoning again?

File: test/dynamo/test_hooks.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        self.assertEqual(x0.grad, x1.grad)
+
+        eager_out = mod(x0)
+        eager_out[0].backward(torch.ones(4))
+
+        comp_out = comp_mod(x1)
+
+        self.assertEqual(cnts.frame_count, 2)
+        comp_out[0].backward(torch.ones(4))
+        self.assertEqual(x0.grad, x1.grad)
```
Comment (ezyang) [inline test/dynamo/test_hooks.py:341]: Re testing, my recommendation is to grep for register_hook calls in test_autograd.py and adapt those tests into Dynamo tests. (Ideally, we could just Dynamo the original test cases themselves but this may imply yaks.) The idea is that test_autograd.py has all the torture cases for hooks we've accumulated over the years that you wanna check.

File: torch/_dynamo/variables/functions.py [hunk 11]
```diff
@@ -110,7 +110,7 @@ def __init__(self, fn, is_constant=False, **kwargs):
             self.is_constant = False
 
         assert isinstance(
-            fn, (types.FunctionType, torch.jit.ScriptFunction)
+            fn, (types.FunctionType, torch.jit.ScriptFunction, functools.partial)
```
Comment (ezyang) [inline torch/_dynamo/variables/functions.py:113]: functools.partial are different from functions, and in all likelihood will need to be modeled by its own VariableTracker subclass. See https://docs.python.org/3/library/functools.html#partial-objects

File: torch/_dynamo/variables/misc.py [hunk 12]
```diff
@@ -790,6 +791,14 @@ def wraps(fn):
             return variables.lists.DequeVariable(
                 items, mutable_local=MutableLocal(), **options
             )
+        elif self.value is functools.partial:
+            guards = options.get("guards", set())
+            guards.add(self.source.make_guard(GuardBuilder.TYPE_MATCH))
```
Comment (ezyang) [inline torch/_dynamo/variables/misc.py:796]: Is this kosher? Won't you unpredictably mutate options["guards"] that was passed in as an argument?

File: torch/_dynamo/side_effects.py [hunk 13]
```diff
@@ -375,6 +380,73 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            #
+            # register_hook on a tensor, AKA backward hooks, have slightly nuanced differences in how they are implemented
+            # when it comes to hooks on objects with sources (inputs, params) vs objects without sources (intermediaries).
+            #
+            # For tensors with a source, we bypass direct inclusion of register_hook calls in the graph.
+            # Instead, these are tracked and stashed as a global variable, enabling their association with tensors in
+            # the residuals. During dynamo's frame creation, these hooks are invoked seamlessly on known reconstructible/fetch-able
+            # tensors. Because a source indicates knowledge of this object outside the torch compile region, and
+            # because we are running residuals firmly before .backward() can be run, it is sound to invoke
+            # `register_hook` on a known tensor.
```
Comment (ezyang) [inline torch/_dynamo/side_effects.py:402]: It's worth noting that you MUST NOT put register_hook calls in the graph for inputs. The canonical example is when the input tensor which had a registered hook is used differentiably outside of the context. The hook must be installed so that it triggers in that case too

File: torch/_dynamo/side_effects.py [hunk 14]
```diff
@@ -375,6 +380,73 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            #
+            # For tensors without a source, the behavior is similar to the above case, except instead of doing the register_hook
+            # call in residuals, we write a register_hook call to the graph. As fx does not understand function
+            # argument, we manually lift the argument up to an input, and use that to register hooks.
```
Comment (ezyang) [inline torch/_dynamo/side_effects.py:406]: I can't tell what this means from the English, specifically "manually lift the argument up to an input"

File: torch/_dynamo/side_effects.py [hunk 15]
```diff
@@ -375,6 +380,73 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            # - For tensors with sources:
+            #   - In the "side_effects" phase of codegen, we iterate over tensors with hooks to:
+            #     - Generate the tensor.
+            #     - Issue a register_hook call on the tensor, linking to the globally stored function.
+            #     - Incorporate a handle if one was established in the eager phase.
+            #  - For tensors without sources:
+            #    - We don't generate any instructions for registering a hook.
+            #    - We lift the fn up as an input.
+            #    - We then manually insert a register_hook call into the graph.
+            # - The handle's exact user-specified name, "last_seen_name", is discerned and associated during STORE_FAST.
```
Comment (ezyang) [inline torch/_dynamo/side_effects.py:425]: This seems very brittle? Like, what if it's seen at two names?

File: torch/_dynamo/symbolic_convert.py [hunk 16]
```diff
@@ -788,7 +792,15 @@ def LOAD_DEREF(self, inst):
...
-        self.symbolic_locals[inst.argval] = self.pop()
+        loaded_vt = self.pop()
+        name = inst.argval
+        # If the last top VT in the stack (just popped) is a handle (see [On tensor.register_hook]), we associate
+        # the stored name.
+        if isinstance(loaded_vt, RemovableHandleVariable):
+            new_name = self.output.new_var(name)
+            new_rhv = loaded_vt.clone(last_seen_name=new_name)
+            loaded_vt = self.replace_all(loaded_vt, new_rhv)
+        self.symbolic_locals[name] = loaded_vt
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:803]: No, just no. Figure out another way to do this.

File: torch/_dynamo/symbolic_convert.py [hunk 17]
```diff
@@ -1861,6 +1875,28 @@ def store_global_weakref(self, name, value):
         if name not in self.output.global_scope:
             self.output.install_global(name, weakref.ref(value))
 
+    def store_hook(self, name, value):
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:1878]: docs?

File: torch/_dynamo/variables/tensor.py [hunk 18]
```diff
@@ -644,6 +645,57 @@ def redistribute_fn_with_prim_types(x):
...
+        elif name == "register_hook":
+            # see [On tensor.register_hook]
+            assert len(args) == 1
+            fn_var = args[0]
+
+            if isinstance(fn_var, variables.NestedUserFunctionVariable):
+                # NestedUserFunctionVariable don't carry their fn, but reconstruction builds it
+                # This should not be onerous to support when needed.
+                unimplemented("NYI - lambda variables as hooks")
+            elif isinstance(fn_var.fn, functools.partial):
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:657]: This seems sus, don't you need to test what kind of VariableTracker you have before you try to access fields on it

File: torch/_dynamo/variables/tensor.py [hunk 19]
```diff
@@ -644,6 +645,57 @@ def redistribute_fn_with_prim_types(x):
...
+                fn = fn_var.fn
+                name = fn_var.fn.func.__name__
+            else:
+                fn = fn_var.fn
+                name = fn_var.fn.__name__
+
+            handle = self.as_proxy().node.meta["example_value"].register_hook(fn)
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:664]: You are unpredictably calling register_hook on either a fake or real tensor. If it's a fake tensor, this is doing nothing, because we are never going to backwards trough it. Is that intended?

File: torch/_dynamo/variables/misc.py (hunk 12 above)
Comment (jansel) [reply torch/_dynamo/variables/misc.py:796]: This is definitely not enough guards.  The correct way to handle this is to recrusively guard on:
- value.func
- value.args
- value.keywords

**Review by jansel — CHANGES_REQUESTED** [review]
The `functools.partial` handling here isn't right.  We should basically just desugar `x: functools.partial` into:
```
lambda *args, *kwargs: x.func(*x.args, *args, **x.keywords, **kwargs)
```
So we have seperate variable builders for x.func, x.args, x.keywords.

Let's put the functools.partial stuff in a different PR with its own tests.

File: test/dynamo/test_hooks.py (hunk 10 above)
Comment (voznesenskym) [reply test/dynamo/test_hooks.py:341]: This is a great suggestion

File: torch/_dynamo/output_graph.py (hunk 8 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:1143]: I think invoking free_symbols on a function type doesn't work :( this is to circumvent the other types that can now flow through here. Much sadness, really.

File: torch/_dynamo/side_effects.py (hunk 13 above)
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:402]: Yea, agreed :)

File: torch/_dynamo/side_effects.py (hunk 14 above)
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:406]: What would you prefer to see this written as?

File: torch/_dynamo/side_effects.py (hunk 15 above)
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:425]: It can only be seen as one name at a time, right? What does the python code look like if it associates it with 2 names?

File: torch/_dynamo/side_effects.py [hunk 20]
```diff
@@ -375,6 +380,73 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            if tensor.source:
+                cg(tensor)
+                cg.extend_output([cg.create_load_attr("register_hook")])
+                cg(hook)
+                cg.extend_output(create_call_function(1, True))
+                if hasattr(handle, "last_seen_name") and handle.last_seen_name:
+                    # register_hook stored with variable name assigned to the handle
+                    cg.extend_output(
+                        [create_instruction("STORE_FAST", argval=handle.last_seen_name)]
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:434]: This is a great point, I should go back to the drawing board on this.

File: torch/_dynamo/symbolic_convert.py (hunk 16 above)
Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:803]: This is temporary, I'll hide this somewhere better (like `loaded_vt.rename(tx, name)`)

File: torch/_dynamo/symbolic_convert.py (hunk 17 above)
Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:1878]: eventually

File: torch/_dynamo/variables/functions.py (hunk 11 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/functions.py:113]: I seem to get along fine modeling them as Function VTs that do a little extra at call_function time...

File: torch/_dynamo/variables/misc.py (hunk 12 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/misc.py:796]: Yeah we should probably use regular old propagation, good catch

File: torch/_dynamo/variables/misc.py [hunk 21]
```diff
@@ -790,6 +791,14 @@ def wraps(fn):
...
+            return variables.UserDefinedClassVariable(
+                self.value,
+                source=self.source,
+                guards=guards,
+            ).call_function(tx, args, kwargs)
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/variables/misc.py:801]: This is not an instance of functools partial, this is calling a ctor of functools.partial

File: torch/_dynamo/variables/tensor.py (hunk 18 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:657]: We can assume it one of the func ones, but yea, lets pop an assert in here

File: torch/_dynamo/variables/tensor.py (hunk 19 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:664]: We just need a way to get the handle out, to recover the name. Super open to feedback on how to do this better tho.

File: torch/_functorch/aot_autograd.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:749]: Yeah, we could come up with some other way I guess, but this makes it an easy stepping stone towards @jansel's plan of not specializing on function ids in favor of just specializing on position + count.

File: torch/_dynamo/output_graph.py (hunk 8 above)
Comment (ezyang) [reply torch/_dynamo/output_graph.py:1143]: `free_symbols` is supposed to work for everything that is an argument type. So if we add new argument types, we need to teach it to ignore it.

But this leads to the other thing I'm pretty unsure about, which is I don't... really want these functions as inputs to the graph.

File: torch/_dynamo/variables/functions.py (hunk 11 above)
Comment (ezyang) [reply torch/_dynamo/variables/functions.py:113]: I'm just saying it's inaccurate and if someone eventually needs to make, e.g., type() or partial.func to work, they will have to undo what you did here

File: torch/_dynamo/variables/misc.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/variables/misc.py:801]: Same deal. If I a class that has no special handling, it should just be a UDCV, right?

File: torch/_dynamo/side_effects.py (hunk 14 above)
Comment (ezyang) [reply torch/_dynamo/side_effects.py:406]: No, as in, I literally do not know what this means.

File: torch/_dynamo/side_effects.py (hunk 15 above)
Comment (ezyang) [reply torch/_dynamo/side_effects.py:425]:
```
h = x.register_hook(func)
h2 = h
```

now the hook shows up at h and h2.

File: torch/_dynamo/variables/tensor.py (hunk 19 above)
Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:664]: But you don't need a handle. You know that `register_hook` is always going to create a fresh handle and so you just make a fresh handle VT. You don't need an actual handle on the inside, and it's better not to have one because it isn't going to do anything useful...

**Comment by voznesenskym** [comment]
> The `functools.partial` handling here isn't right. We should basically just desugar `x: functools.partial` into:
> 
> ```
> lambda *args, *kwargs: x.func(*x.args, *args, **x.keywords, **kwargs)
> ```
> 
> So we have seperate variable builders for x.func, x.args, x.keywords.
> 
> Let's put the functools.partial stuff in a different PR with its own tests.

https://github.com/pytorch/pytorch/pull/108846/files

File: torch/_dynamo/variables/tensor.py (hunk 19 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:673]: good point

File: torch/_dynamo/side_effects.py (hunk 15 above)
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:425]: I think that will just work due to how we handle h2 = h, but let's not guess, I will add a test.

File: torch/_higher_order_ops/invoke.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def invoke(fn, grad):
+    return invoke_op(fn, grad)
+
+
+invoke_op = HigherOrderOperator("invoke")
+
+
+@invoke_op.py_impl(ProxyTorchDispatchMode)
```
Comment (bdhirsh) [inline torch/_higher_order_ops/invoke.py:21]: Pretty cool though :) Is this is our way of (eventually) hiding backward hook from AOTAutograd so that we don't effectively allow_in_graph them, and instead let dynamo inspect them during compiled backwards?

File: torch/_functorch/aot_autograd.py (hunk 9 above)
Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:749]: Is there a discussion somewhere I can read on that?

It's not clear to me why making hook fns inputs into the FX graph will help you avoid specializing on fn ID's. If a user changes the hook that they register onto an intermediate, won't we still need to recompile, which would require guarding on the id of the actual hook fn?

e.g.:
```
@torch.compile
def fn_with_custom_hook(inp, hook):
    intermed = inp * 2
    intermed.register_hook(hook)
    out = intermed * 2
    return out
    
inp = torch.randn(4, requires_grad=True)
out1 = fn_with_custom_hook(inp, hook1)
# This should probably recompile?
out2 = fn_with_custom_hook(inp, hook2)
```

File: torch/_higher_order_ops/invoke.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+def inner_trace(fn, grad):
+    mode = _get_current_dispatch_mode()
+    if isinstance(fn, functools.partial):
+        fn.__name__ = fn.func.__name__  # type: ignore[attr-defined]
+    grad = torch._functorch.aot_autograd.from_fun(grad)
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_higher_order_ops/invoke.py:34]: Perfect

File: torch/_functorch/aot_autograd.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:749]: Discussed offline

File: test/dynamo/test_hooks.py (hunk 10 above)
Comment (voznesenskym) [reply test/dynamo/test_hooks.py:557]: This became a huge yack shave. But what I can do is port some of the more interesting tests to here.

File: torch/_dynamo/side_effects.py (hunk 14 above)
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:412]: todo: reword

File: torch/fx/experimental/symbolic_shapes.py [hunk 24]
```diff
@@ -216,6 +216,8 @@ def free_symbols(val: Union[SymInt, torch.Tensor]) -> Set[sympy.Symbol]:
         for s in val:
             r |= free_symbols(s)
         return r
+    elif isinstance(val, functools.partial):
+        return set()
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:220]: this is suspicious. Because functools.partial can contain arbitrary objects, which might themselves have free symbols.

File: torch/_higher_order_ops/invoke.py [hunk 25]
```diff
@@ -0,0 +1,88 @@
+import functools
```
Comment (ezyang) [inline torch/_higher_order_ops/invoke.py:1]: We can land this in its own PR, with its own tests.

File: torch/_higher_order_ops/invoke.py [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+        "call_function", fn, proxy_args, {}, name="invocation"
+    )
+    grad = original_fn(grad)
+    grad = track_tensor_tree(grad, out_proxy, constant=None, tracer=mode.tracer)
+    return torch._functorch.aot_autograd.to_fun(grad)
+
+
+@invoke_op.py_impl(FakeTensorMode)
+def inner_fake(fn, grad, reenter):
+    return fn(grad)
```
Comment (ezyang) [inline torch/_higher_order_ops/invoke.py:60]: But you can't do this, because fn might have Python side effects, no? It would be better to default to NOT calling the inner func for meta prop, rather than unsoundly YOLO calling it

Comment (voznesenskym) [reply torch/_higher_order_ops/invoke.py:60]: Hmm, true, I think for meta prop we can try speculating it and make assurances about output shape. Let me pull this into its own PR.

File: torch/_higher_order_ops/invoke.py (hunk 25 above)
Comment (voznesenskym) [reply torch/_higher_order_ops/invoke.py:1]: SGTM. Bikeshedding, I hate this name...

File: test/dynamo/test_hooks.py (hunk 10 above)
Comment (ezyang) [reply test/dynamo/test_hooks.py:594]: that's fine too

File: torch/fx/experimental/symbolic_shapes.py [hunk 27]
```diff
@@ -207,19 +207,31 @@ def free_symbols(val: Union[SymInt, torch.Tensor]) -> Set[sympy.Symbol]:
...
-    elif isinstance(val, (tuple, list)):
+    elif isinstance(val, (tuple, list, set)):
         r = set()
         for s in val:
             r |= free_symbols(s)
         return r
+    elif isinstance(val, dict):
+        r = set()
+        r |= free_symbols(set(val.keys()))
+        r |= free_symbols(list(val.values()))
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:226]: we shouldn't need to actually materialize the set/list here; maybe we just need to support `free_symbols` on an iterator.

I'm a bit suspicious about all of the new types supported here though, originally this basically let you cover FX node argument types, but now it is a bit more expansive than that.

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/pull/109537/files
